### PR TITLE
Introduce `Logger.experiment_dir`

### LINKED ISSUE #14188: Introduce `Logger.experiment_dir`
## 🚀 Feature

Introduce a property/method `experiment_dir` on the abstract logger interface. All loggers must implement this and return a path to where their files are being saved. 

### Motivation

The Trainer currently determines where checkpoints are saved like so:
```py
os.path.join(logger.save_dir, str(logger.name), logger.version, "checkpoints")
```

As you can see, it assumes that there is a rigid folder structure `save_dir/name/version`. Every logger has to follow this structure.
However, at least one logger (wandb) struggles with this, because it has an additional concept of "project" that does not fit well within this layout (https://github.com/Lightning-AI/lightning/issues/14054). Furthermore, the "name" may change AFTER the files were written, so there is a disconnect between the folder structure and what is represented "online" in the wandb ui. 

One observation is that the trainer actually never cares about any of the subfolders. All it needs is the full path. This is why I believe, if we let the logger decide how to assemble the path, we will avoid many issues and disagreements with 3rd party experiment management. 

We are struggling with making all loggers agree to a standardized layout, so maybe we shouldn't try to do that, at least not in every aspect of how the loggers work (https://github.com/Lightning-AI/lightning/issues/12028#issuecomment-1088325894). 

### Pitch

Introduce

```py
def experiment_dir(self) -> str:
    """Returns the (relative) path to where logs and artifacts get saved for the current version."""

```

This should probably be a relative path, so that the Trainer can prepend a root dir if needed (default cwd). 
An implementation of a logger could then look like this:

```py
def experiment_dir(self):
    return os.path.join(self.save_dir, self.name, self.version)
```

And the checkpointing in the Trainer would save files to

```py
os.path.join(logger.experiment_dir(), "checkpoints") 
```
for example.

### Implications

1. After introducing experiment dir, we could consider deprecating `save_dir` and `log_dir` or remove it from the base interface.
2. The code here in `Trainer.log_dir` would probably get simplified: https://github.com/Lightning-AI/lightning/blob/fe9e5d55bf7991ba36b76d6adae9075b93dfcaa0/src/pytorch_lightning/trainer/trainer.py#L2208-L2219
3. The code in the checkpoint callback would get simplified too: https://github.com/Lightning-AI/lightning/blob/fe9e5d55bf7991ba36b76d6adae9075b93dfcaa0/src/pytorch_lightning/callbacks/model_checkpoint.py#L595-L608
4. It would immediately address these issues, as the LightningCLI can write to the experiment dir without ambiguity: #14162, #12748

### Alternatives

Struggle

### Additional context

Discussions in:
https://github.com/Lightning-AI/lightning/issues/14054

Current blockers: 
#12177
______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/Lightning-AI/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/Lightning-AI/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/Lightning-AI/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/Lightning-AI/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging PyTorch Lightning, Transformers, and Hydra.

cc @borda @tchaton @justusschock @awaelchli @edward-io @ananthsub @rohitgr7 @kamil-kaczmarek @Raalsky @Blaizzy

### PR DESCRIPTION
**Pull request by tshu-w**
"## What does this PR do?

Introduce the property first.

Part of #14188

Next steps:
1. Change `Trainer.log_dir` to `experiment_dir` and `ModelCheckpoint` Implementation #14640,
2. Update implementations of some Loggers, and maybe drop some unnecessary properties(`name`, `version`, etc)
3. Introduce `Trainer.experiment_dir` and deprecate `Trainer.log_dir`

### Does your PR introduce any breaking changes? If yes, please list them.

When the `Logger.save_dir` is `None` (DummyLogger, CometLogger(online), Deprecated LoggerCollection, etc), the `Trainer.log_dir` is `None` either, and the checkpoints directory is `default_root_dir/name/version`.
https://github.com/Lightning-AI/lightning/blob/b33e3ef18aa8f50dd905e3762aaaa3117aa65705/src/pytorch_lightning/callbacks/model_checkpoint.py#L594-L604

Since there are many Loggers that can have a `save_dir` of None, I prefer to make `logger.experiment_dir` return a relative path. 
cc @awaelchli 

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃
"

### CODE REVIEW

**Comment by carmocca** [comment]
If this is going to be sequenced so the properties are introduced first, and then used. I would like that you open the follow-up PR already so that we can see that things worked as expected (CI passed).

**Comment by tshu-w** [comment]
> If this is going to be sequenced so the properties are introduced first, and then used. I would like that you open the follow-up PR already so that we can see that things worked as expected (CI passed).

Sure, I can open the following PR first. Should I make this to draft PR first.

**Comment by justusschock** [comment]
Closed in favor of #14640
